q2_quality_control/_threshold_graph/_visualizer.py

```diff
@@ -48,33 +48,26 @@ def decontam_score_viz(output_dir, decontam_scores: qiime2.Metadata,
                        weighted: bool = True, bin_size: float = 0.02):
     _check_inputs(**locals())
     df = decontam_scores.to_dataframe()
-    temp_values = df['p'].tolist()
-    temp_values = np.array(temp_values)
-    if np.count_nonzero(np.isnan(temp_values)) == len(temp_values):
-        raise ValueError("All input p values are NA")
-
-    table_trans = table.transpose()
-    temp = table_trans.sum(axis='columns')
-    read_nums = np.array(temp.tolist())
-
-    contam_asvs = 0
-    true_asvs = 0
-    contam_reads = 0
-    true_reads = 0
-    index = 0
-    values = []
-    true_read_nums = []
-    for val in temp_values:
-        if np.isnan(val) is False:
-            values.append(val)
-            true_read_nums.append(read_nums[index])
-            if val < threshold:
-                contam_asvs = contam_asvs + 1
-                contam_reads = contam_reads + read_nums[index]
-            else:
-                true_asvs = true_asvs + 1
-                true_reads = true_reads + read_nums[index]
-        index = index + 1
+    if df['p'].isna().all():
+        raise ValueError("No p-values exist for the data provided.")
+
+    read_nums = table.sum(axis='rows')
+
+    p_vals = df['p'].dropna()
+    filt_read_nums = read_nums[p_vals.index]
+
+    contams = (p_vals < threshold)
+
+    contam_asvs = contams.sum()
+    true_asvs = len(contams) - contam_asvs
+    unknown_asvs = len(df['p']) - true_asvs - contam_asvs
+    percent_asvs = contam_asvs / (contam_asvs + true_asvs) * 100
+
+    contam_reads = filt_read_nums[contams[contams].index].sum()
+    true_reads = filt_read_nums.sum() - contam_reads
+    unknown_reads = read_nums.sum() - true_reads - contam_reads
+    percent_reads = contam_reads / (contam_reads + true_reads) * 100
+
```

Comment:
These changes are just vectorizing the operations using pandas, which already makes these kinds of summary statistics easy to calculate.

I also added `unknown` as a category to the report, as it seems relevant to include now that `p-values` can be nan.

Reply:
Okay awesome, it looks like the `unknown` is being reported in the table below the graph?